### PR TITLE
Proposal: Modification of Receive Split proposal.

### PR DESCRIPTION
**Pull request by bwplotka**
👋🏽 @Chans321 @brancz @GiedriusS @kakkoyun @squat @yashrsharma44 @metalmatze 

Sorry for not paying attention to the past proposal, but when we revisited and tried to review  https://github.com/thanos-io/thanos/pull/3845 and https://github.com/thanos-io/thanos/pull/3580  with @kakkoyun we found the potentially easier idea that might be more cleaner for users and maintainers. Feel free to give feedback!

First of all, thanks for all of your hard work on this, and experimental implementations in PR 🤗 We only want to propose different configuration really, but the idea stays exactly the same.

Some learnings for the future:

1. Let's not merge things prematurely. The previous proposal was missing some important parts:
   * Goals were stating implementation goals, not the usage goals. We should focus on usage goals to avoid focusing too much on implementation and not seeing other options.
   * Proposal was missing Alternatives.
   * Motivation was motivating the implementation. I think it should rather motivate GOALS, not implementation yet.
   * Motivation was not clear. Statements like `Right now the receiver handles ingestion, routing and writing, thus leading to too many responsibilities in a single process.` might need some proof.
   * There were some left overs like `-Additionally, we touch possibility for adding a consistent hashing mechanism and buffering the incoming requests.` What that means?
   * The related ticket was for something else - for different database reconfiguration @squat proposed. Given that there was not enough context for this proposal.
   * Lack of migration guide for such breaking change of already stable receive.

2. Implementation PR led to confusions. 
  * It was too large, ideally we can split work into smaller pieces. It was hard to review properly, agree on all items, not miss important piece etc.
  * It was marked as almost ready, which led @yashrsharma44 to spend time on rebasing and cleaning all of it. Instead, it was missing many elements like documentation, migration docs, mixings, dashboards, alerts that should be at least well planned ahead. If we would merge any of those two in this state it could led to even bigger confusion.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: docs/proposals/202012_receive_split.md
Comment (brancz) [inline docs/proposals/202012_receive_split.md:94]: If this is the only con, then there are simple solutions to solve this.

**Comment by brancz** [comment]
I don't see sufficient reason here to change the direction of this project entirely when it's so close to the finish line, and two maintainers having signed off on the previous architecture.

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [inline docs/proposals/202012_receive_split.md:28]: I think this is still suggesting the implementation.

```
* Reduce the downtime of ingestion part of Thanos
```

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [inline docs/proposals/202012_receive_split.md:63]: I just inherited this from the previous proposal. Any ideas what was that about?

File: docs/proposals/202012_receive_split.md
Comment (kakkoyun) [inline docs/proposals/202012_receive_split.md:58]: 👍

File: docs/proposals/202012_receive_split.md
Comment (kakkoyun) [reply docs/proposals/202012_receive_split.md:63]: cc @Chans321 @squat @brancz

**Review by kakkoyun — DISMISSED** [review]
Looks great. With this new proposed implementation, it draws more clear migration path and it's backwards compatible. It would much easier to roll these out.

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [reply docs/proposals/202012_receive_split.md:94]: This is left overall, sorry, will be more explicit

Comment (bwplotka) [reply docs/proposals/202012_receive_split.md]: I clarified, could you double check?

**Comment by bwplotka** [comment]
> I don't see sufficient reason here to change the direction of this project entirely when it's so close to the finish line, and two maintainers having signed off on the previous architecture. Also there is a docs PR out already.

I understand your @brancz  feelings and it's entirely our bad we did not participate in this discussion, however. It's my fault. I would still love if you could take a look on 3 aspects:

1. > change the direction of this project entirely 

Is it entirely changing? From user there is no much different. Indeed it changing a lot of implementation wise - it avoids tons of work which is good?

2. > when it's so close to the finish line

I would argue with this. I don't think it's close for proper usage by users. Also it's creating breaking changes which violates the stability guarantee we made with receive. Of course we are not in 1.0 but I would expect some migration plan for receive useage. I mentioned reasoning in https://github.com/thanos-io/thanos/pull/4036#issue-612348300 point 2.

This proposal here allows to avoid tricky migration.

3. > , and two maintainers having signed off on the previous architecture

What's the difference if 1, two or hundred of maintainers signed off if the alternative we propose here was not considered (at least in the view of previously written proposal)?

4. I explained the rationales in Proposal. The need of single-process receive and route you commented on is arguable and really low priority issue. The main issue is the user confusion of architectural design and each components purpose, boileplate of another command, inconsistency with Querier pattern. I would love to at least look and consider In my opinion easier solution.

**Comment by kakkoyun** [comment]
@brancz @bwplotka I also want to see this push through the finish line as quickly as possible. If could just tweak the proposal and modify the existing implementation, it'd be faster to merge this. We can even make this happen for v0.20.0 by pushing the cut of RC for a week (we just released a major version recently anyways). I'm happy to help with the effort. 

Current implementation changes the interface of the existing commands drastically, there is no easy migration path using current implementation. Maybe we should acknowledge the proposed design and implementation drift apart. 

FWIW, either we can proposal or implementation, I want to have these changes on the main branch. Initial benchmarks by @metalmatze proved that there are nice gains in here.

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [inline docs/proposals/202012_receive_split.md:58]:
```suggestion
3. This is more future proof for potential advanced cases like chain of routers -> receives -> routers -> receives for federated writes, so trees with depth n.
```

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:17]: Should it be `receives` or `receiver`? I am getting confused :(
```suggestion
to another receiver based on hashring. This allows setting optional deployment model were only Routing Receives are using hashring files and does the routing and replication. That allows ingesting Receivers to not handle any routing or hashring, only receiving multi tenant writes.
```

Comment (yashrsharma44) [reply docs/proposals/202012_receive_split.md:17]: Because out here in this line - `That allows ingesting Receivers to not handle any routing or hashring, only receiving multi tenant writes.`, we use `receiver` :)

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:66]: So by **_not using a flag_** we enforce a certain behaviour, shouldn't we use a flag to enforce what we want?

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:28]:
```suggestion
* Reduce downtime of the ingestion operation of Thanos Receiver
```

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:57]:
```suggestion
2. Allow consistent pattern with Query. We don't have separate StoreAPI component for proxying, we have that baked into Querier. This has been proven to be flexible and understandable, so I would like to propose similar pattern in Receive.
```

**Review by yashrsharma44 — COMMENTED** [review]
Some nitpick and few questions from my side.
Otherwise LGTM 🥇

**Comment by brancz** [comment]
I'm not necessarily arguing with the technological direction (I do agree this is a good, maybe better direction), but I find it worrisome that we let parts of the community work on something for a long time only to step in and "know it better" and turn over a lot of work someone has put into something. This is being everything else but welcoming to new contributors, and I've noticed this behavior more than once. Part of letting people grow is allowing them to make mistakes and not stepping in and having to be part of every conversation.

Clearly, there are enough maintainers to back this proposal up from a technical perspective, so I'm not going to block this proposal, but for the record, I want to state that I find this behavior damaging for the Thanos project as a whole. Not everyone can always be part of every conversation and we need to accept that, even if it means technological decisions we may not necessarily agree with may be made.

**Comment by bwplotka** [comment]
Thanks, Frederic, this is important to all of us. As we discussed offline, there were lots of offline chats and even meeting between @kakkoyun, @Chans321 @yashrsharma44. The trigger for this proposal was actually done on pre-last Contributor Thanos Hours, which is also recorded and available here: 

```
Topic: Thanos Contributor Office Hours
Date: Apr 1, 2021 11:00 AM Universal Time UTC

Meeting Recording:
https://zoom.us/rec/share/VoV2AVCrZYmvaPk27kSftXnTAtVEb-OQqkEiYVxfXNLFWfdSu_sNiAr7CL92888Z.seIHezcDVw2szx8f
```

Unfortunately, you as the mentor of @Chans321 were not updated on all of those communications, sorry for that!

It's true we should be welcoming, but it's also worth to mention we started this work in Community Bridge Q4 2020. Lot's of new ideas came after that I think are fair to consider.

> Not everyone can always be part of every conversation and we need to accept that, even if it means technological decisions may be made that an individual may not necessarily agree with.

Yea, but I treat this proposal as an improvement over the past work, all reusing the previous work and foundations.

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [reply docs/proposals/202012_receive_split.md]: You are right: It should be "Receiver": https://thanos.io/tip/contributing/contributing.md/#components-naming

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [reply docs/proposals/202012_receive_split.md]: Hm, wonder if operation is a right word 🤔

Comment (bwplotka) [reply docs/proposals/202012_receive_split.md]: changed to `logic`

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [reply docs/proposals/202012_receive_split.md]: Looks like no idea anymore. Removing then.

File: docs/proposals/202012_receive_split.md
Comment (bwplotka) [reply docs/proposals/202012_receive_split.md]: Well that's the thing. If you don't use hashring flag, it's by design impossible to do hashring forwarding, because you did not specify hashring, right? (:  That's why we can use it as quite explicit turn off of forwarding logic, no?

**Review by bwplotka — COMMENTED** [review]
Updated, please PTAL again @yashrsharma44 @kakkoyun 

Thanks for good review @yashrsharma44 !

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:74]:
```suggestion
This potentially makes the receiver more difficult to operate and understand for Thanos users. I would argue this is however much harder in overall Thanos deployment. Otherwise, this option is exactly the same.
```

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:61]: Not sure about this suggestion, just that the old statement sounded contrapositive 😆  
```suggestion
* Receiver with `--receive.hashrings` forwards or replicate requests, without this, it would route the request straight to multi-tsdb.
```

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:62]: Similar to above
```suggestion
* Receiver with ` --receiver.local-endpoint` would create storage for multi TSDB, without this flag, we will skip creating any resources for multi TSDB.
```

**Review by yashrsharma44 — COMMENTED** [review]
LGTM 🔥 , just small questions about wording of the statement 😁

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [inline docs/proposals/202012_receive_split.md:72]:
```suggestion
1. Split the Receiver component into receive-route and Receiver (and ensure ease of resharding events).
2. Evaluate any effects on performance by simulating scenarios and collecting and analyzing metrics.
3. Use consistent hashing to avoid reshuffling time series after resharding events. The exact consistent hashing mechanism to be used needs some further research.
4. Migration: We document how the new architecture can be set up to have the same general deployment of the old architecture. (We run router and Receiver on the same node).
```

**Review by yashrsharma44 — DISMISSED** [review]
Resolved the issues after Office Hours, looks good to me!

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [reply docs/proposals/202012_receive_split.md:62]: Resolved during office hours :)

File: docs/proposals/202012_receive_split.md
Comment (yashrsharma44) [reply docs/proposals/202012_receive_split.md:61]: Resolved during office hours :)

File: docs/proposals/202012_receive_split.md
Comment (GiedriusS) [inline docs/proposals/202012_receive_split.md:41]: Grammar nit:
```suggestion
We can call this mode a "Routing Receiver". Similarly, we can skip specifying any hashring to Thanos Receiver (`--receive.hashrings-file=<path>`), explicitly purposing it only for ingesting. We can call this mode "Ingesting Receiver".
```

**Review by GiedriusS — DISMISSED** [review]
LGTM modulo that one grammar nit. It indeed sucks a lot that we have some initial implementations of the previous version but at the same time I feel that it won't be too much work to update them for this improved version as we are only changing how this new mode will be turned on, right?

**Comment by bwplotka** [comment]
Thank you all for feedback and reviews!

**Pull request merged**
